### PR TITLE
warn in `sim.epsilon` if too many grid cells

### PR DESCRIPTION
**Pull request by tylerflex**

### CODE REVIEW

File: tidy3d/components/simulation.py
Comment (shashwat-sh) [inline tidy3d/components/simulation.py:53]: This is about 650 cells in each direction - just curious as to how this number was chosen? Assuming a double precision epsilon array where each entry is 8 bytes, I guess this number leads to about 2.2GB RAM - or 4.4GB for complex numbers. Was this the main consideration?

**Review by shashwat-sh — APPROVED** [review]

File: tidy3d/components/simulation.py
Comment (tylerflex) [reply tidy3d/components/simulation.py:53]: This was just chosen empirically through playing around with the tests and picking the number where sim.epsilon took a second or so to run.

Comment (shashwat-sh) [reply tidy3d/components/simulation.py:53]: Ok, thanks.

**Comment by tylerflex** [comment]
I think the test might be failing because of the memory requirement issue. 
I might try the following: make the lower limit for warning a function of `num_grid_cells` times `number of structures`. Because we'd need to loop this function over structures anyway. That way the tests can pass and it will give a better estimate for cases with smaller sims but more structures.

**Comment by momchil-flex** [comment]
> I think the test might be failing because of the memory requirement issue. I might try the following: make the lower limit for warning a function of `num_grid_cells` times `number of structures`. Because we'd need to loop this function over structures anyway. That way the tests can pass and it will give a better estimate for cases with smaller sims but more structures.

This makes more sense indeed, as the same number of cells can take vastly different time depending on the number of structures. I think scaling should be linear with both once my inside improvements are merged in. You may want to hold off testing until then?

**Comment by momchil-flex** [comment]
Actually, scaling may be linear already, just with a much worse prefactor.

**Comment by tylerflex** [comment]
Ok.  I already made these changes so we can either 
1. merge yours first and then adjust the warning lower bound.
2. merge this and then adjust the lower bound in your PR after you rebase.
Either way, I'll leave this open

**Comment by tylerflex** [comment]
looks like tests pass now, so i think we can merge this now or after your geometry improvements are in

**Comment by momchil-flex** [comment]
Let's merge and then I'll play around with that test in my PR. But how about changing the message a bit to simply

```
                f"Simulation contains {int(np.prod(grid.num_cells)):.2e} grid cells and "
                f"{len(self.structures):.2e} structures. Epsilon calculation may be slow."
```

**Pull request merged**
